Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp

```diff
@@ -0,0 +1,419 @@
+{
+    // TODO: visit attributes
+    m_stringBuilder.append(m_indent, "struct ", structDecl.name(), " {\n");
+    {
+        IndentationScope scope(m_indent);
+        for (auto& member : structDecl.members()) {
+            m_stringBuilder.append(m_indent);
+            visit(member.type());
+            m_stringBuilder.append(" ", member.name());
+            for (auto &attribute : member.attributes()) {
+                m_stringBuilder.append(" ");
+                visit(attribute);
+            }
+            m_stringBuilder.append(";\n");
+        }
+    }
+    m_stringBuilder.append(m_indent, "};\n\n");
+}
+
+void FunctionDefinitionWriter::visit(AST::VariableDecl& variableDecl)
+{
+    ASSERT(variableDecl.maybeTypeDecl());
+
+    visit(*variableDecl.maybeTypeDecl());
+    m_stringBuilder.append(" ", variableDecl.name());
+    if (variableDecl.maybeInitializer()) {
+        m_stringBuilder.append(" = ");
+        visit(*variableDecl.maybeInitializer());
+    }
+}
+
+void FunctionDefinitionWriter::visit(AST::Attribute& attribute)
+{
+    AST::Visitor::visit(attribute);
+}
+
+void FunctionDefinitionWriter::visit(AST::BuiltinAttribute& builtin)
+{
+    if (builtin.name() == "vertex_index"_s) {
+        m_stringBuilder.append("[[vertex_id]]");
+        return;
+    }
+
+    if (builtin.name() == "position"_s) {
+        m_stringBuilder.append("[[position]]");
+        return;
+    }
+
+    ASSERT_NOT_REACHED();
+}
+
+void FunctionDefinitionWriter::visit(AST::StageAttribute& stage)
+{
+    switch (stage.stage()) {
+    case AST::StageAttribute::Stage::Vertex:
+        m_stringBuilder.append("vertex");
+        break;
+    case AST::StageAttribute::Stage::Fragment:
+        m_stringBuilder.append("fragment");
+        break;
+    case AST::StageAttribute::Stage::Compute:
+        m_stringBuilder.append("compute");
+        break;
+    }
+}
+
+void FunctionDefinitionWriter::visit(AST::LocationAttribute& location)
+{
+    m_stringBuilder.append("[[attribute(", location.location(), ")]]");
+}
+
+void FunctionDefinitionWriter::visit(AST::TypeDecl& type)
+{
+    AST::Visitor::visit(type);
+}
+
+void FunctionDefinitionWriter::visit(AST::ArrayType& type)
+{
+    ASSERT(type.maybeElementType());
+    ASSERT(type.maybeElementCount());
+    m_stringBuilder.append("array<");
+    visit(*type.maybeElementType());
+    m_stringBuilder.append(", ");
+    visit(*type.maybeElementCount());
+    m_stringBuilder.append(">");
+}
+
+void FunctionDefinitionWriter::visit(AST::NamedType& type)
+{
+    if (type.name() == "i32"_s)
+        m_stringBuilder.append("int");
+    else if (type.name() == "f32"_s)
+        m_stringBuilder.append("float");
+    else if (type.name() == "u32"_s)
+        m_stringBuilder.append("unsigned");
+    else
+        m_stringBuilder.append(type.name());
+}
+
+void FunctionDefinitionWriter::visit(AST::ParameterizedType& type)
+{
+    switch (type.base()) {
+    case AST::ParameterizedType::Base::Vec2:
```

Comment:
Very curious about your opinions on having `Vec2` be explicit AST nodes, rather than just regular types.

On one hand, the WGSL grammar explicitly mentions them (I think because they take `<>` parameters, and the editors want special handling for that), but on the other hand, if we handle all types systemically, it removes the need for special casing them throughout our code.